### PR TITLE
[PiranhaJava] Delete statements using EasyMock and JUnit API

### PR DESCRIPTION
**Pull request by ketkarameya**
Added new matchers for the EasyMock and Junit API

### CODE REVIEW

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java [hunk 1]
```diff
@@ -369,9 +363,9 @@ private API getXPAPI(
     for (PiranhaMethodRecord methodRecord : methodRecordsForName) {
       // when argumentIndex is specified, if mit's argument at argIndex doesn't match xpFlagName,
       // skip to next method property map
-      Optional<Integer> optionalArgumentIdx = methodRecord.getArgumentIdx();
+      OptionalInt optionalArgumentIdx = methodRecord.getArgumentIdx();
```
Comment (ketkarameya) [inline java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:366]: My type change tool suggested this :)

File: java/piranha/src/main/java/com/uber/piranha/config/MethodRecord.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import static com.uber.piranha.config.PiranhaRecord.getValueBooleanFromMap;
+import static com.uber.piranha.config.PiranhaRecord.getValueStringFromMap;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+import javax.annotation.Nullable;
+
+/** A class representing a method configuration record from properties.json */
+public class MethodRecord {
```
Comment (ketkarameya) [inline java/piranha/src/main/java/com/uber/piranha/config/MethodRecord.java:26]: I extracted the super class MethodRecord from PiranhaMethodRecord

File: java/piranha/config/properties.json [hunk 3]
```diff
@@ -38,6 +38,32 @@
       "argumentIndex": 0
     }
   ],
+  "unnecessaryTestMethodProperties" : [
+    {
+      "methodName": "when",
+      "argumentIndex": 0,
+      "recieverType":"org.mockito.Mockito",
```
Comment (mkr-plse) [inline java/piranha/config/properties.json:45]: typo: `receiverType`

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java [hunk 4]
```diff
@@ -705,30 +699,38 @@ private Description updateCode(
...
-      endPos = state.getEndPosition(stmt);
-      return SuggestedFix.builder().delete(stmt);
+
+    if (enclosingMit != null
+        && enclosingEst != null
+        && config
+            .getUnnecessaryTestMethodRecords()
+            .stream()
+            .map(
+                mthd ->
```
Comment (mkr-plse) [inline java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:723]: expand `mthd` to `method`?

File: java/piranha/src/main/java/com/uber/piranha/config/PiranhaRecord.java [hunk 5]
```diff
@@ -53,4 +53,21 @@ static Integer getArgumentIndexFromMap(Map<String, Object> map, String argumentI
...
+   * string.
+   *
+   * @param map - map corresponding to a method property
+   * @param key - key to check the corresponding value
+   * @return the corresponding boolean value
+   */
+  @Nullable
+  static Boolean getValueBooleanFromMap(Map<String, Object> map, String key) {
+    Object value = map.get(key);
+    if (value instanceof String && !value.equals("")) {
```
Comment (mkr-plse) [inline java/piranha/src/main/java/com/uber/piranha/config/PiranhaRecord.java:68]: `!value.isEmpty()`

File: java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java [hunk 6]
```diff
@@ -541,6 +545,27 @@ public void testRemoveSpecificAPIpatterns() throws IOException {
             "  when(experimentation.isToggleEnabled(\"OTHER_FLAG\")).then(ans);",
             "  when(foobar()).thenReturn(false);",
             "  when(foobar()).thenAnswer(ans);",
+            "// Easymock Test scenarios",
```
Comment (mkr-plse) [inline java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:548]: Can we split these into multiple tests (easymock, ...)?

**Review by mkr-plse — COMMENTED** [review]
Can we also update the README to reflect the new property?

File: java/README.md [hunk 7]
```diff
@@ -135,6 +135,14 @@ Additionally, usages of `STALE_FLAG` will be removed as if the enum itself had b
 
 Finally, the setting `linkURL` in the properties file is to provide a URL describing the Piranha tooling and any custom configurations associated with the codebase. 
 
+Another top-level optional field is `unnecessaryTestMethodProperties`. Within that, there is an array of JSON objects, 
+having the fields `methodName`, `argumentIndex`, `receiverType`, `returnType` and `isStatic`.
+
+In practice, often the refactored code produced by `piranha` results in unnecessary method invocations. For instance,
```
Comment (mkr-plse) [inline java/README.md:141]: let us get rid of the term `often` :). How about "The refactored code may contain unnecessary method invocations".

File: java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java [hunk 8]
```diff
@@ -500,6 +500,203 @@ public void testIgnoresPrefixMatchFlag() throws IOException {
...
+            "  Answer ans = new Answer<Integer>() {\n"
+                + "      public Integer answer(InvocationOnMock invocation) throws Throwable {\n"
+                + "        return (Integer) invocation.getArguments()[0];\n"
+                + "      }};",
+            "",
+            "",
+            "",
+            "",
+            "",
+            "  boolean b1 = someWrapper(when(true).thenCallRealMethod());",
```
Comment (mkr-plse) [inline java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:568]: Is this wrapping common?

**Review by mkr-plse — APPROVED** [review]
LGTM

File: java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java (hunk 8 above)
Comment (ketkarameya) [reply java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:568]: I don't know, TBH. I just added it as a corner case to the test suite, assuming that if such a scenario appears it should not be refactored because the method `someWrapper` could be producing some required side effect.

Comment (mkr-plse) [reply java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:568]: no worries. this is fine. we can iterate on this subsequently.
